### DISCUSSION TITLE
Corrupt audio/video recordings with multiple AudioCapture nodes

### DISCUSSION THREAD

**Discussion by msarvestani** [bonsai-rx/bonsai#1390] [Q&A]
Hi all. I have a simple workflow to record the audio/video streams from a couple of webcams. For the purposes of debugging, I'm not worried about synchronization between cameras. I'm having a funky issue when more than one AudioCapture node is enabled.

I can record audio/video from one camera just fine. But when I try to record from two cameras, the files are corrupt. I narrowed the problem down to having a second AudioCapture node enabled. This second node reads from the microphone of the second camera and isn't just a duplicate of the first AudioCapture node. For some reason, having this second node enabled disrupts recordings from the first camera ... The disrupted recordings are shorter than expected and the audio contains noise. 

Also, when I look at the output of the two different AudioCapture nodes, they seem to be reading some combination of both microphones, even when I try to isolate sound input to one or the other....which is confusing.

Any help would be greatly appreciated.
![image](https://github.com/bonsai-rx/bonsai/assets/12864873/d18f29b6-db44-4d46-bb48-0be40d404e13)

[Webcams_debug.zip](https://github.com/bonsai-rx/bonsai/files/11593351/Webcams_debug.zip)

**Comment by bruno-f-cruz** [comment]
Hey!

You have a bunch of race conditions that are likely subject to CPU load (likely to increase when adding a second camera). For instance, the first WithLatestFrom is effectively checking what is the latest buffer available form the mic and returns it. If, for some reason, you delay the update of this "latest" buffer, you will likely get double buffers in our Wav file.

To start debugging this problem, I would record the audio and the video independently (without the reactive logic) and try to align posthoc. If the problem is still there, then I agree that it might be some sort of hardware/drivers bug.

Let us know what you find!
B

**Reply by msarvestani** [reply]
Thanks, makes sense that it might be a load issue. 

I removed logicals and can reproduce the issue with the most minimal workflow involving just two AudioCapture nodes.  I'll see if I can get some other microphone peripherals and figure out if it's specific to the logitech webcams I'm using.

![image](https://github.com/bonsai-rx/bonsai/assets/12864873/dbc55abb-0291-404f-9b23-b28f72a257dd)

**Reply by msarvestani** [reply]
Update: I've tried 2 different PCs and different pairs of microphones like apple headphones, USB mic's, etc. All reproduce the same issue where audio recordings are corrupt if I have more than 1 AudioCapture node enabled. I also reinstalled Bonsai, and same issue. My CPU/Memory usage are okay and not maxed out while Bonsai is running.

Any advice on what else I can try?

**Reply by bruno-f-cruz** [reply]
Can you share the minimum workflow that reproduces the issue? In the meanwhile, can you use MatrixWriter to save the audio and check if the problem persists?
Thanks!

**Reply by msarvestani** [reply]
[minimal_audio_capture.zip](https://github.com/bonsai-rx/bonsai/files/11605484/minimal_audio_capture.zip)
I actually started with MatrixWriter/Videowriter(ffmpeg) and thought the issue was with ffmpeg code. But it seems unrelated to how I capture. 

Thanks.

**Reply by bruno-f-cruz** [reply]
Just to clarify, have you tried using MatrixWriter for the audio as well?

**Reply by msarvestani** [reply]
I hadn't tried just with audio but I just did this test (attached), and it's the same. Bonsai also crashes if I have the second AudioCapture node enabled and using MatrixWriter. 
[minimal_audio_capture_MatrixWriter.zip](https://github.com/bonsai-rx/bonsai/files/11605663/minimal_audio_capture_MatrixWriter.zip)

**Reply by bruno-f-cruz** [reply]
Just to cover all basis, can you try to save the file as a flat binary file and open it in python for instance?. Just to make sure it is not a problem with the pipe/ffmpeg string.

**Reply by msarvestani** [reply]
Ok, same thing, assuming I saved to binary correct. I saved the binary file from MatrixWriter like this: 
![image](https://github.com/bonsai-rx/bonsai/assets/12864873/8096afcc-8696-4052-a606-d844d3d42571)

and produced two files: one with second audiocap node disabled, one with it enabled.

consistent with the other capture methods, the recording with the second audiocap enabled is shorter, even though the timer is fixed.

here is a plot of the two binary file readouts:
![binary_readout](https://github.com/bonsai-rx/bonsai/assets/12864873/7f9968e2-9aa4-460b-8062-030dcb2bb8f8)
